Assignment4-Beighlor M

bluebandit95 opened the pull request.

bluebandit95 committed:
Create Assignment1.yml

bluebandit95 committed:
Create Azure Login with OIDC.yml

bluebandit95 committed:
Update Assignment1.yml

bluebandit95 committed:
Assignment1.yml

bluebandit95 committed:
Assignment1.yml

bluebandit95 committed:
Assignment1.yml

bluebandit95 committed:
Assignment1.yml

bluebandit95 committed:
Azure Login with OIDC.yml

bluebandit95 committed:
Create Assignment2.ps1

bluebandit95 committed:
Add files via upload

bluebandit95 committed:
Merge pull request /Assignment3-Beighlor-M

Assignment 3

bluebandit95 committed:
Assignment3.ps1

bluebandit95 committed:
Merge branch 'IntelliTect-Samples:main' into main

bluebandit95 committed:
Assignment 4.ps1

jenny-curry commented:
Can you add me as a contributor on your subscription? I noticed you are using a different subscription than in previous assignments, if it's not in the EWU Azure tenant, I'll need your tenant id value to be submitted as well. 
![image](https://github.com/IntelliTect-Samples/EWU-CSCD396-2023-Fall/assets/43355843/a4b433f2-4ed7-4488-bb15-1880ed927c7e)

bluebandit95 commented:
Hi, I added you as a contributor. I had added your work email instead of your ewu one, sorry about that. I removed your work email and added your ewu one. 

Here is my tenant id:
![image](https://github.com/IntelliTect-Samples/EWU-CSCD396-2023-Fall/assets/97564858/963b2c67-6e77-4327-8b5c-8cea69956653)

jenny-curry commented:
1. Create Log Analytics Workspace and Custom Table
   https://learn.microsoft.com/en-us/azure/azure-monitor/logs/quick-create-workspace?tabs=azure-powershell
   https://learn.microsoft.com/en-us/cli/azure/monitor/log-analytics/workspace/table?view=azure-cli-latest#az-monitor-log-analytics-workspace-table-create

- Log Analytics Workspace Created ✅
- Create Custom Table with columns 'TimeGenerated' (DateTime type)and 'LogLevel' (string type)✅

2. Create a Data Collection Rule and Endpoint
   https://learn.microsoft.com/en-us/azure/azure-monitor/logs/set-up-logs-ingestion-api-prerequisites

- Data Collection Endpoint Created ✅
- Custom Table is Created with Data Collection Rule ✅ (Using the portal, recreation is required to use DCR on the Custom Table)
- Data Collection Rule Created ✅


3. Create Function App
   https://learn.microsoft.com/en-us/azure/azure-functions/create-first-function-vs-code-powershell#publish-the-project-to-azure
   https://learn.microsoft.com/en-us/azure/azure-functions/functions-monitor-log-analytics?tabs=csharp#setting-up
   https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=python-v2%2Cin-process%2Cfunctionsv2&pivots=programming-language-powershell

- Function App Created ✅
  - Use [this tutorial](https://learn.microsoft.com/en-us/azure/azure-functions/create-first-function-vs-code-powershell) to setup your function app
- Configured Function App Diagnostic setting to the previously created log analytics workspace ✅
- Enable System Assigned identity on Function App ✅
- Add Az.Accounts as a package install by editing the requirements.psd1 app file - https://learn.microsoft.com/en-us/azure/azure-functions/functions-reference-powershell?tabs=portal#dependency-management 

  - Copy SampleResources/Assn4FunctionProj/SendLogs/run.ps1 file contents to your function. Replace the values in the curly braces. The Assn4FunctionProj directory is an example step 1 of [creating your local functions project](https://learn.microsoft.com/en-us/azure/azure-functions/create-first-function-vs-code-powershell#create-an-azure-functions-project)
  - Replace table name, DCE uri, and DCR Immutable Id values in run.ps1 with your values. 
- Created PowerShell Function with Http Inbound and Outbound Bindings 
  - See SampleResources/Assn4FunctionProj/SendLogs/function.json for sample
- Assigned 'Monitoring Metrics Publisher' and 'Contributor' role to the function app managed identity over the scope of your resource group. This identity will be executing the PowerShell function. https://learn.microsoft.com/en-us/azure/azure-monitor/logs/tutorial-logs-ingestion-portal#assign-permissions-to-the-dcr ✅
- Executed the function through PowerShell commands: 
      $Date = Get-Date ([datetime]::UtcNow) -Format O 
      $Body = @{LogLevel='Error';TimeGenerated=$Date} | ConvertTo-Json
      Invoke-RestMethod -Method POST -Uri 'https://<Fn App Name>.azurewebsites.net/api/<Function Name>?code=<My Function API Key here>' -Body <Body> 
  - See monitor tab on the function to view requests. 
  - If this is a success, your log should be present in your custom table
- Log Found in your Created Custom Table ✅

4. Create API Management Instance ✅
   https://learn.microsoft.com/en-us/azure/api-management/get-started-create-service-instance-cli
- Import Function App as an API in your API Management Instance ✅
 https://learn.microsoft.com/en-us/azure/api-management/import-function-app-as-api 


5. Create a PowerShell script called Assignment4.ps1 on your branch within the Assignment4 folder ✅
